js/server/modules/@arangodb/testutils/aql-graph-traversal-generic-tests.js

```diff
@@ -2962,6 +2963,117 @@ function testSmallCircleKShortestPathEnabledWeightCheckWithMultipleLimitsGen(tes
   });
 }
 
+function testSmallCircleFilterOptimization(testGraph) {
+  // Which Graph is used here does not matter, enough to test with one of the graphs
+  // we picked the smallCircle graph at random
+  assertTrue(testGraph.name().startsWith(protoGraphs.smallCircle.name()));
+
+  const optimizableConditions = [
+    `v.color == "blue"`,
+    `"blue" == e.color`,
+    `p.edges[1].color == "blue"`,
+    `"blue" == p.vertices[1].color`,
+    `p.edges[*].color ALL == "blue"`,
+    `p.vertices[*].color ALL == "blue"`
+  ];
+
+  // Optimize a single condition
+  const filtersToTest = optimizableConditions.map(f => `FILTER ${f}`);
+  // More complex test.
+  for (let first = 0; first < optimizableConditions.length; ++first) {
+    for (let second = first + 1; second < optimizableConditions.length; ++second) {
+      // Optimize two filter nodes, and erase the filter nodes
+      filtersToTest.push(`FILTER ${optimizableConditions[first]} FILTER ${optimizableConditions[second]}`);
+    }
+  }
+
+  for (const filterCondition of filtersToTest) {
+    const query = `
+      FOR v,e,p IN 1..3 OUTBOUND "${testGraph.vertex('A')}" GRAPH ${testGraph.name()}
+        ${filterCondition}
+        return v
+    `;
+    const plan = AQL_EXPLAIN(query, {});
+
+
+    assertEqual(findExecutionNodes(plan, "FilterNode").length, 0, query + " Still has FilterNode");
+
+    const traversals = findExecutionNodes(plan, "TraversalNode");
+    assertEqual(traversals.length, 1, query + " Somehow we have removed the Traversal");
+    const travNode = traversals[0];
+    assertTrue(travNode.hasOwnProperty("condition"), JSON.stringify(travNode.condition));
+  }
+
+  // Non optimizable filters
+  const nonOptimizableFilters = [
+    `DOCUMENT(v.docId).color == "blue"`,
+    `DOCUMENT(e.docId).color == "blue"`,
+    /* ASSSERT has side-effects on false so it should not be optimized */
+    `ASSERT(v.color == "blue", "Color is not blue")`,
+    `DATE_YEAR(p.vertices[1].timestamp) == DATE_YEAR(DATE_ADD(p.vertices[0].timestamp, 1, "y"))`,
```

Comment:
Is there a reason why this particular filter condition cannot be optimized? DATE_YEAR and DATE_ADD should be deterministic and side-effect-free...

Reply:
The issue here is that the path access is on both sides of the equation.
For now we only optimize if the other side does not have access to v , e or p.
This can be improved.